### ISSUE TITLE
Unexpected circularity error when indirectly observing value in `while` loop

### ISSUE THREAD

**Issue by p-szm** [microsoft/TypeScript#50037]
# Bug Report

So, this is a weird one and apologies for a convoluted repro (I couldn't simplify it any further).

### 🔎 Search Terms

Inference, any, unions, nested objects

### 🕗 Version & Regression Information

- This changed between versions 4.6.4 and 4.7.4
- This is broken on nightly

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-beta#code/CYUwxgNghgTiAEYD2A7AzgF3kgRgKwC54BveAMwEsQJgiBlJAWxAFUULV4BfAbgFgAUIIwBPAA4IGzNhxTwAvPBmoAsiEY4QMAIzwAPkvar1mmACZ+QgRRQYtZKGATKUajVt3FB8H-FESiACJtQMtfeAA3KAgAVxAiFBj3GEsuQUEbOxgHJ0NZN1MzEm9ff3j4QLNQkp8o2PLE5NT06zJ4AApcPAA6SmpgbrKFeUVgwIBKYoFwgHcACwoIBHaMGDjJr2nw32R0LDq4hWx8XqoaboOQMO2fAHpb8IA9AH4a7d3MSOi4osVL65uHywjCgYiIpAA2gBrEAiBJJUwAXXhyW4R2IvDe4RBYghlzMiKO2h48Hu8AAwkxmLY-As0NgYlgoChgLSEFoYEgYPAZosIPBgBQ0KCJLA3mkBBKgA)

### 💻 Code

```ts
declare const obj: { field: SomeUnion };

type SomeUnion = UnionMember1 | UnionMember2;

interface UnionMember1 {
    type: "1";
    value: number;
}

interface UnionMember2 {
    type: "2";
    value: number;
}

if (obj.field.type === "1") {
    while (true) {
        const value = obj.field.value;
        //    ^?
        const value2 = value;
        const map: { [key: number]: number } = {};
        map[value2] = 1; // Comment this out and the error will disappear
    }
}
```

### 🙁 Actual behavior

`value` is of type `any`, unless the line `map[value2] = 1` is commented out.

### 🙂 Expected behavior

`value` should be of type `number`

**Comment by nesk** [comment]
Here's another reproducible example if needed, it's shorter but the behavior is really similar to the code provided with the issue.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=9#code/JYOwLgpgTgZghgYwgAgApwOajmYB7EZAbwFgAoZZECADzAH4AuZAVQCUAZcgX3IBMICADZwoKGAFcQCXAWQAHUQGcIAOVpgWUIQAoxS+QRXM2EA0YgBKZuiwgc+EOXJCIYKhq1Dm7DsgA+yFICMKAQfMgAvB4A7qycOgBEABZgYPKMAPSZtHAAtvKuAHQIeHmZQsBKDIoYEJEAjImW5DHJwK7IOtR0XpbE5JSlINXI+oYjKNFwMXDA7jBuCMndntotFMjDSnjFQngYemYTKkUAVjsgOpYblD2a2lEKympruuMWlkX3PM5ktIYoO4iMhuMhssgACrtJTIKrIMB4ZBwIT7OISJSgDAIvDyAC0rgAbhAhMjZvNyEA)

### 💻 Code

```ts
interface Pagination {
  next?: URL
}
declare function parseNextUrl(response: Response): Pagination

let nextUrl: URL | undefined = new URL("http://example.com/list?page=1")
while (nextUrl) {
  const response = await fetch(nextUrl)
  console.log(response.json())
  nextUrl = parseNextUrl(response).next
}

export { } // This is to allow using top-level await
```

**Comment by NWYLZW** [comment]
Looks like Promise cause

[Playground](https://www.typescriptlang.org/play?target=9#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwApkYIAueAVQCUAZASkoAUYcBbLAZxAB5a6APgCwAKDEQQGeKhAAPDFTKUB8AD7w0oRFlnB4AXhkgA7tXpEARAAsMGAA7kA9E-lR29yQDowHJxG4MAH57KABzEAMARksGMTETayxJeCJZBSUIBngAbzF4eF9ULmlSCEN4KBMoLGlEQjT5RTI40QKXAoKAPSD84wyyCrKxAF940Xl7HBhpHPgx0SA)

```ts
declare function foo(url: URL): Promise<URL>

let nextUrl: URL | undefined = new URL("http://example.com/list?page=1")

while (nextUrl) {
  const url = await foo(nextUrl)
  //    ^?
  nextUrl = url
}

export { }
```

**Comment by NWYLZW** [comment]
Simplified code for this issue

[Playground](https://www.typescriptlang.org/play?target=9#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwApkYIAueAVQCUAZASkoG8BYAKHnlRAA8NKtOhwC+HDhBAZufDFTKD68AD7w0oRFh7B4AXhkB3avSIAiABYYMAB3IB6O3ygBba5IB0YHM7sQsAZwwAfmsoAHMQXQBGUwZxdgNzLEl4Ih5+eQgGeDZOeC9UQLUyPQJidLkyBncKji4HLi4APSC6mQyS-VIIUQ4gA)

```ts
declare function foo(url: URL): {
  next: URL
}

let nextUrl: URL | undefined = new URL("http://example.com/list?page=1")

while (nextUrl) {
  const url = foo(nextUrl).next
  //    ^?
  nextUrl = url
}
```
